File: backend/src/routes/hotels.ts
import express from 'express';
import { getHotelProvider } from '../services/inventory';
import { HotelQuerySchema } from '../services/inventory/types';

const router = express.Router();

router.get('/search', async (req, res) => {
  try {
    const {
      destination,
      checkIn,
      checkOut,
      adults = '2',
      children = '0',
      minPrice,
      maxPrice,
      currency = 'USD',
      sort = 'relevance',
    } = req.query;

    if (!destination || !checkIn || !checkOut) {
      return res.status(400).json({
        error: 'Missing required parameters: destination, checkIn, checkOut',
      });
    }

    const provider = getHotelProvider();
    const query = HotelQuerySchema.parse({
      destination: destination as string,
      checkIn: checkIn as string,
      checkOut: checkOut as string,
      adults: parseInt(adults as string),
      children: parseInt(children as string),
      minPrice: minPrice ? parseInt(minPrice as string) : undefined,
      maxPrice: maxPrice ? parseInt(maxPrice as string) : undefined,
      currency: currency as string,
      sort: sort as string,
    });

    const hotels = await provider.search(query);

    return res.json({
      success: true,
      count: hotels.length,
      hotels,
    });
  } catch (error: any) {
    console.error('Hotel search error:', error);
    return res.status(500).json({
      error: error.message || 'Failed to search hotels',
    });
  }
});

export default router;

File: backend-python/app/routes/hotels.py
"""Hotels routes — search via SerpApi."""

from fastapi import APIRouter, HTTPException, Query

from app.services.serpapi_provider import serpapi_provider

router = APIRouter()


@router.get("/search")
async def search_hotels(
    destination: str = Query(...),
    checkIn: str = Query(...),
    checkOut: str = Query(...),
    adults: int = 2,
    children: int = 0,
    minPrice: int | None = None,
    maxPrice: int | None = None,
    currency: str = "USD",
    sort: str = "relevance",
):
    hotels = await serpapi_provider.search_hotels(
        destination=destination,
        check_in=checkIn,
        check_out=checkOut,
        adults=adults,
        children=children,
        min_price=minPrice,
        max_price=maxPrice,
        currency=currency,
        sort=sort,
    )
    return {"success": True, "count": len(hotels), "hotels": hotels}
